src/nonlinear.jl

```diff
@@ -1,148 +1,134 @@
-    return logpdf(prob.noise, [obs_t] .- z_t)
-end
+    n1 = noise(prob.obs_noise, 1)
+    n = Vector{typeof(n1)}(undef, T+1) # Latent noise
+    n[1] = n1
 
-function _likelihood(
-    prob::StateSpaceProblem,
-    alg::ConditionalGaussian,
-    z_t,
-    obs_t,
-    noiseType
-)
-    # If given a noise distribution, provide the log pdf w.r.t. the
-    # distribution.
-    throw(ArgumentError("No implementation for prob = StateSpaceProblem, alg = ConditionalGaussian, and noiseType = $noiseType"))
+    z1 = prob.h(u[1], prob.params, 1) + noise(prob.obs_noise, 1)
+    z = Vector{typeof(z1)}(undef, T+1) # Observables generated
+    z[1] = z1
+
+    # Simulate it, homie
+    for t in 2:T+1
+        n[t] = noise(prob.noise, t)
+        u[t] = prob.f(u[t-1], prob.params, t-1) .+ prob.g(u[t-1], prob.params, t-1) * n[t]
+        z[t] = prob.h(u[t], prob.params, t) .+ noise(prob.obs_noise, t)
+    end
+
+    return StateSpaceSolution(z, u, n, nothing)
 end
 
-"""
-    _solve(prob::StateSpaceProblem, params)
-
-Returns a `StateSpaceSolution` using the provided parameters `params`.
-Calculates the likelihood if the likelihood function is available.
-"""
-function _solve(
-    prob::StateSpaceProblem{isinplace, dType, uType,tType,noiseType,oType}, 
-    alg::ConditionalGaussian, 
-    params
-) where {isinplace, dType, uType,tType,noiseType,oType <: Nothing} # No observables
-    t0, T = prob.tspan
-
-    # Initial noise
-    z0 = prob.h(prob.u0, params, 1)
-
-    # Preallocate arrays
-    u = Vector{typeof(prob.u0)}(undef, T+1)
-    z = Vector{typeof(z0)}(undef, T+1)
-    n = noisetype(prob) <: Vector ? 
-        Vector{eltype(prob.noise)}(undef, T+1) :
-        Vector{typeof(prob.u0)}(undef, T+1) # Store sampled transition noise
+function CommonSolve.solve!(
+    prob::StateSpaceProblem{isinplace, ftype, gtype, htype, wtype, vtype, utype, ttype, otype}, 
+    args...; 
+    kwargs...
+) where {isinplace, ftype, gtype, htype, wtype, vtype, utype, ttype, otype}
+    # Preallocate values
+    T = prob.tspan[2]
 
+    u = Vector{utype}(undef, T+1) # Latent states
     u[1] = prob.u0
-    z[1] = z0
+    
+    n1 = noise(prob.obs_noise, 1)
+    n = Vector{typeof(n1)}(undef, T+1) # Latent noise
+    n[1] = n1
 
+    z1 = prob.h(u[1], prob.params, 1) + noise(prob.obs_noise, 1)
+    z = Vector{typeof(z1)}(undef, T+1) # Observables generated
+    z[1] = z1
+
+    @info "" n1 n
+
+    # Simulate it, homie
+    loglik = 0.0
     for t in 2:T+1
-        n[t-1] = latent_noise(prob, t-1)
-        u[t] = prob.f(u[t - 1], params, t-1) + prob.g(u[t - 1], params, t-1) .* n[t-1] # Should this be n[t] or n[t-1]?
-        z[t] = prob.h(u[t], params, t)
+        n[t] = noise(prob.noise, t)
+        u[t] = prob.f(u[t-1], prob.params, t-1) .+ prob.g(u[t-1], prob.params, t-1) * n[t]
+        z[t] = prob.h(u[t], prob.params, t) .+ noise(prob.obs_noise, 1)
+        err = z[t] - prob.observables[t]
+        loglik += loglikelihood(err, prob.obs_noise, t)
```

Comment:
hmm, are there different solvers for this kind of thing?

Reply:
Yeah, I believe there's the Kalman filter for linear forms, and various extensions that handle nonlinear models. I think.